### ISSUE TITLE
-doxygen encounters syntax error in input when last member of class has doxygen comment afterwards

### ISSUE THREAD

**Issue by gieseanw** [swig/swig#1636]
When `-doxygen` is enabled, we encounter an error

 > Syntax error in input(3)

for code like this:
```
struct Foo
{
  int a; //! some value
};
```

Or
```
struct Foo
{
  int a; //! some value
  int b; //! another value
};
```

But not this:

```
struct Foo
{
  int a; //! some value
  int b;
};
```

**Comment by huckw** [comment]
I saw the same behavior.  It occurred in a file I didn't need docs for and so I just modified the comment from `//! my comment` to `// ! my comment`

**Comment by mcfarljm** [comment]
The form being used here is not actually a doxygen comment.  If you change to:
```
struct Foo
{
  int a; //!< some value
};
```
then it will work.  See http://www.doxygen.nl/manual/docblocks.html#memberdoc.

I'm not 100% sure whether the syntax error from swig is intended, but it seems reasonable.  At a minimum, changing to correct doxygen syntax is a workaround.

**Comment by HenningJ** [comment]
The same thing happens with [Doxygen member groups](http://www.doxygen.nl/manual/grouping.html#memgroup):
```
struct Foo
{
  ///@{ 
  int a;
  ///@}
};
```
(which is correct doxygen syntax)

**Comment by baoilleach** [comment]
For anyone arriving here looking for a workaround, this problem is 'fixed' by adding an empty statement just before the end of the class definition, e.g. 

```
  ; // Added to workaround https://github.com/swig/swig/issues/1636
}; // end of class
```

**Comment by ojwb** [comment]
The original example here is not really correct:

```
struct Foo
{
  int a; //! some value
  int b; //! another value
};
```

A `//!` comment documents what *follows*, so this has no doxygen comment for `a`, the comment `some value` for `b` and then an extra comment `another value` which doesn't get attached to anything - if you run Doxygen on this example you get a warning due to this:

```
test.h:3: warning: Member a (variable) of struct Foo is not documented.
```

If you check the generated documentation, then `some value` is attached to `b` and `another value` doesn't appear.

So really the input here is wrong and should be fixed to use `//!<` comments which document what comes before them (as @mcfarljm suggested) or the `//!` comments moved before the things they document, and then the documentation comments get attached to the intended members by both Doxygen and SWIG.

However SWIG shouldn't really fail with an opaque parse error for this case, and I think the example with member groups is valid Doxygen markup.  I think it just needs an extra rule in the parser which hopefully won't introduce a conflict.  I'll take a look.

**Comment by ojwb** [comment]
#1609 is the same issue but for `enum` rather than `class`.

**Comment by wsfulton** [comment]
All fixed in 8a24c19
